Import-Package and other requirements not translated as Maven

merks opened the pull request:
dependencies when generating pom

Reduce the number of MavenMappings.

Use the new CBI aggregators features described here to map based Maven
metadata in p2 and to build Maven dependencies from java.package
requirements:

https://github.com/eclipse-cbi/p2repo-aggregator/issues/11

Disable the validation repositories because they're not needed given the
platform repository itself contains all requirements, and keeping them
also results analysis anomalies because Orbit IUs are resolved where
direct-from-Maven IUs are actually in the repository.

https://github.com/eclipse-platform/eclipse.platform.releng/issues/67

merks commented:
Note that this requires the latest tools from https://download.eclipse.org/cbi/updates/p2-aggregator/tools/nightly/latest to try it in the IDE.   Installing these will allow to edit both the resources in their structured editors:

![image](https://user-images.githubusercontent.com/208716/178284873-9ee2b4f6-2104-4c1f-a834-b270106d3298.png)

I used the analyzer to ensure that every dependency maps to an artifact and that every version range resolves to an actual version that actually exists in Maven.  The view gives an overview of all the generated poms.  Here you can see how a package import is mapped to a dependency:

![image](https://user-images.githubusercontent.com/208716/178286436-f1d33032-9feb-4da6-a381-c55338f94064.png)

The log will contain information about potential problems or cases where a mapping is no longer needed.

merks commented on publish-to-maven-central/SDK4Mvn.aggr:
```diff
@@ -37,37 +37,30 @@
-  <mavenMappings namePattern="org.w3c.dom.events" groupId="org.eclipse.orbit.bundles" artifactId="org.w3c.dom.events"/>
-  <mavenMappings namePattern="org.w3c.dom.smil" groupId="org.eclipse.orbit.bundles" artifactId="org.w3c.dom.smil"/>
-  <mavenMappings namePattern="org.w3c.dom.svg" groupId="org.eclipse.orbit.bundles" artifactId="org.w3c.dom.svg"/>
+  <mavenMappings namePattern=".*" groupId="\$maven-groupId\$" artifactId="\$maven-artifactId\$" versionPattern=".*" versionTemplate="\$maven-version\$"/>
```
I asked via email for your feedback in here but none was forthcoming:

https://github.com/eclipse-cbi/p2repo-aggregator/issues/11

I think we/you can live with this explicit opt-in rule.  I want to offer that flexibility to ensure that new behavior is opt-in rather than "we think this is best for you so you get it whether you want it or not."

mickaelistria replied:
> "we think this is best for you so you get it whether you want it or not."

IMO, this usually results in better results. More configuration is more complexity, more room for errors; it is often less productive. I think this is the case here, and I don't see addition of this rule as an added-value over the current state (where this rule is hardcoded as default when possible) but more as an extra complication that doesn't make maintenance of this .aggr file, nor of the aggregator itself, easier.

merks replied:
Sometimes opinions differ.  This option does not and will not require further/future maintenance.

merks commented on publish-to-maven-central/SDK4Mvn.aggr:
```diff
@@ -37,37 +37,30 @@
-  <mavenMappings namePattern="org.w3c.dom.smil" groupId="org.eclipse.orbit.bundles" artifactId="org.w3c.dom.smil"/>
-  <mavenMappings namePattern="org.w3c.dom.svg" groupId="org.eclipse.orbit.bundles" artifactId="org.w3c.dom.svg"/>
+  <mavenMappings namePattern=".*" groupId="\$maven-groupId\$" artifactId="\$maven-artifactId\$" versionPattern=".*" versionTemplate="\$maven-version\$"/>
+  <mavenDependencyMapping iuNamePattern="(?!.*(org\.eclipse\.)).*|org\.eclipse\.emf.*|org\.eclipse\.ecf.*|org\.eclipse\.jetty.*" namespacePattern=".*" namePattern=".*" groupId="!" artifactId="!"/>
+  <mavenDependencyMapping namespacePattern="java\.package" namePattern="(org.eclipse.jdt).internal.(compiler\.(apt|tool))" groupId="$1" artifactId="$1.$2"/>
+  <mavenDependencyMapping namespacePattern="java\.package" namePattern="(org.eclipse.jdt).internal.(compiler\.apt)\..*" groupId="$1" artifactId="$1.$2"/>
+  <mavenDependencyMapping namespacePattern="java\.package" namePattern="(org.eclipse.jdt)(.internal|.core)?.compiler.*" groupId="$1" artifactId="$1.core"/>
+  <mavenDependencyMapping namespacePattern="java\.package" namePattern=".*" groupId="*" artifactId="*" versionRangePattern="(.*)"/>
```
This (mapping of package requirements and requirements in general) was described here:

https://github.com/eclipse-cbi/p2repo-aggregator/issues/11#issuecomment-1180208947

But I already drew attention to that in the comment of this PR.

merks commented:
@akurtakov 

This version I've checked with the aggregator analysis editor that all generated Maven dependencies resolve to ones that exist in Maven.  The ones that don't are marked with ** and from the wrapped search you can see only the new never-before-published-to-maven plugin is marked that way:

![image](https://user-images.githubusercontent.com/208716/187065315-96e8af24-8a5b-45b3-a5af-5340036a4f0b.png)

This additionally generates dependencies based on package imports which seems important given how much bundle requirements are being replaced by package imports.

Publishing should of course use the latest version of the aggregator for the package import support (as well as for the new support to use the Maven coordinates from the p2 metadata):

https://download.eclipse.org/cbi/updates/p2-aggregator/products/nightly/latest

merks commented:
@akurtakov Should I merge this now?

akurtakov commented:
I would defer the decision to @mickaelistria or @sravanlakkimsetti as they spent time in this cycle to investigate publishing.

merks commented:
@sravanlakkimsetti

I'd like to understand the maven publishing process better.  Where can I learn more about how it works?  The steps for EMF publishing to Maven are really quite complicated, and I only publish a small subset...

mickaelistria commented:
There is https://wiki.eclipse.org/Platform-releng/Publish_To_Maven_Central which is more or less up-to-date. The effective entry point is /eclipse.platform.releng/publish-to-maven-central/CBIaggregator.sh , it does contain a bit of documentation inline about the process.

merks commented:
@sravanlakkimsetti 

One concern I have is with the pom enriching logic...

In particular I think this assumes that there is not already a `name` element.

https://github.com/eclipse-platform/eclipse.platform.releng/blob/b25099b5d011009deda219b765f69ea2c6fde657/publish-to-maven-central/src/org/eclipse/platform/releng/maven/pom/ArtifactInfo.java#L66-L71

That method is called from here which assumes there is necessarily a `description` element:

https://github.com/eclipse-platform/eclipse.platform.releng/blob/b25099b5d011009deda219b765f69ea2c6fde657/publish-to-maven-central/src/org/eclipse/platform/releng/maven/pom/EnrichPoms.java#L76-L82

I changed the logic to properly populate the name element and the description element, but plugins generally don't generally have a description so that element would be missing.   I ran into this issue publishing EMF to Maven yesterday where the final Nexus check complained about the description being missing.  I modified the EMF pom enhancer to set the description to the name if the description was missing, but now I've also restored the logic in the aggregator such that if there is no description the name will be used as the description, so the aggregator produces a pom like this:

![image](https://user-images.githubusercontent.com/208716/187887885-37134fcd-b1dd-4d4a-a07c-dbb0d3ee04a2.png)

This is what leads me to be concerned that the current enhancer logic will create two `name` elements.

Note too that org.eclipse.platform.releng.maven.pom.ArtifactInfo is full of assumptions about git.eclipse.org as the SCM.

I suppose it would be better to create a new issue...

mickaelistria commented:
@merks can you please open a dedicated issue about the name/description concerns?

merks commented:
> @merks can you please open a dedicated issue about the name/description concerns?

I’m out of the office the rest of the day with only my iPhone. So not until much later.
